shaders/khr_ray_tracing_extended/raygen.rgen

```diff
@@ -0,0 +1,170 @@
+/* Copyright (c) 2021 Holochip Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 the "License";
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#version 460
+#extension GL_EXT_ray_tracing : enable
+
+layout(binding = 0, set = 0) uniform accelerationStructureEXT topLevelAS;
+layout(binding = 1, set = 0, rgba8) uniform image2D image;
+layout(binding = 2, set = 0) uniform CameraProperties 
+{
+	mat4 viewInverse;
+	mat4 projInverse;
+} cam;
+
+struct Payload
+{
+  vec4 color;
+  vec4 intersection; // {x, y, z, intersectionType}
+  vec4 normal; // {nx, ny, nz, distance}
+};
+
+vec3 heatmap(float value, float minValue, float maxValue)
+{
+  float scaled = (min(max(value, minValue), maxValue) - minValue) / (maxValue - minValue);
+  float r = scaled * (3.14159265359 / 2.);
+  return vec3(sin(r), sin(2 * r), cos(r));
+}
+
+layout(location = 0) rayPayloadEXT Payload hitValue;
+layout (constant_id = 0) const uint render_mode = 0;
+layout (constant_id = 1) const uint maxRays = 12;
+
+void main() 
+{
+	const vec2 pixelCenter = vec2(gl_LaunchIDEXT.xy) + vec2(0.5);
+	const vec2 inUV = pixelCenter/vec2(gl_LaunchSizeEXT.xy);
+	vec2 d = inUV * 2.0 - 1.0;
+
+	vec4 origin = cam.viewInverse * vec4(0,0,0,1);
+	vec4 target = cam.projInverse * vec4(d.x, d.y, 1, 1) ;
+	vec4 direction = cam.viewInverse*vec4(normalize(target.xyz), 0) ;
+
+	float tmin = 0.001;
+	float tmax = 10000.0;
+
+	uint max_rays = maxRays;
+	max_rays = 100;
+	if (render_mode != 0)
+	{
+		max_rays = 1;
+	}
+
+	uint object_type = 100;
+	vec4 color = vec4(0, 0, 0, 0);
+	// 0 = normal, 1 = shadow, 2 = AO
+	uint current_mode = 0;
+	float expectedDistance = -1;
+	for (uint i = 0; i < max_rays && current_mode < 100 && color.a < 0.95 && (color.r < 0.99 || color.b < 0.99 || color.g < 0.99); ++i)
+	{
+		traceRayEXT(topLevelAS, gl_RayFlagsOpaqueEXT, 0xff, 0, 0, 0, origin.xyz, tmin, direction.xyz, tmax, 0);
+		object_type = uint(hitValue.intersection.w);
+		const vec3 object_intersection_pt = hitValue.intersection.xyz;
+		const vec3 object_normal = hitValue.normal.xyz;
+		if (render_mode != 0)
```

Comment:
would it possible to use defines that match enum values for `render_mode` and `object_type`, to improve readability? or maybe comments would be enough?